File: interactive/src/binary-asset.ts
// Shared, reusable loader for the compact binary interactive-data format
// (WP15 §3). Replaces large numeric-array JSON payloads with a small
// versioned JSON manifest (validated like any other activity data, via zod)
// plus one deterministic binary file, fetched with `fetch(...).arrayBuffer()`
// and decoded into typed arrays -- never eval, never innerHTML.
//
// Python encoder: scripts/binary_asset.py (keep both in sync: dtypes,
// 4-byte alignment, little-endian). Activity-specific reconstruction (e.g.
// expanding a neighbour-index matrix back into a logical target-value
// matrix) lives in each activity's own *-data.ts, on top of this module.

import { z } from "zod";
import { resolveDataUrl } from "./urls";

const IDENTIFIER_TOKEN = /(^|_)(id|ids|uid|sub|subject|participant|site|mrn|name|dob)($|_)/i;

export const BINARY_DTYPES = ["float32", "uint16"] as const;
export type BinaryDtype = (typeof BINARY_DTYPES)[number];

const ELEMENT_SIZE: Record<BinaryDtype, number> = { float32: 4, uint16: 2 };

export const binarySectionSchema = z
  .object({
    dtype: z.enum(BINARY_DTYPES),
    shape: z.array(z.number().int().nonnegative()).min(1),
    byteOffset: z.number().int().nonnegative(),
    byteLength: z.number().int().nonnegative(),
  })
  .strict();

export type BinarySection = z.infer<typeof binarySectionSchema>;

export const binaryManifestBaseSchema = z
  .object({
    schemaVersion: z.number().int().positive(),
    activity: z.string().min(1),
    binary: z
      .object({
        path: z.string().min(1),
        byteLength: z.number().int().nonnegative(),
        sha256: z.string().regex(/^[0-9a-f]{64}$/i, "sha256 must be a 64-character hex digest"),
      })
      .passthrough(),
    sections: z.record(binarySectionSchema),
  })
  .passthrough();

export type BinaryManifestBase = z.infer<typeof binaryManifestBaseSchema>;
export type DecodedSection = Float32Array | Uint16Array;

// Avoid duplicate network fetches of the same binary file on one page (e.g.
// a re-render, or two activities that -- in principle -- shared a payload).
const bufferCache = new Map<string, Promise<ArrayBuffer>>();

async function fetchArrayBuffer(url: URL): Promise<ArrayBuffer> {
  const key = url.href;
  let pending = bufferCache.get(key);
  if (!pending) {
    pending = (async () => {
      const res = await fetch(url.href, { credentials: "omit", mode: "same-origin" });
      if (!res.ok) throw new Error(`HTTP ${res.status} for ${url.href}`);
      return res.arrayBuffer();
    })();
    bufferCache.set(key, pending);
  }
  return pending;
}

/** Exposed for tests only -- do not call from application code. */
export function _clearBufferCacheForTests(): void {
  bufferCache.clear();
}

async function sha256Hex(buffer: ArrayBuffer): Promise<string | null> {
  const subtle = typeof crypto !== "undefined" ? crypto.subtle : undefined;
  if (!subtle || typeof subtle.digest !== "function") return null; // no secure-context SubtleCrypto available
  const digest = await subtle.digest("SHA-256", buffer);
  return Array.from(new Uint8Array(digest))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

// TypedArray views always use the host platform's native byte order. Every
// browser engine in practice is little-endian, but this loader is written to
// be CORRECT on a (today, essentially nonexistent) big-endian host too: it
// feature-detects platform endianness once and falls back to an explicit
// little-endian DataView decode rather than silently mis-reading bytes.
const PLATFORM_IS_LITTLE_ENDIAN = new Uint16Array(new Uint8Array([1, 0]).buffer)[0] === 1;

function decodeSection(buffer: ArrayBuffer, name: string, section: BinarySection): DecodedSection {
  const itemSize = ELEMENT_SIZE[section.dtype];
  if (section.byteOffset % itemSize !== 0) {
    throw new Error(`section "${name}": byteOffset ${section.byteOffset} is not aligned to its ${section.dtype} element size`);
  }
  if (section.byteLength % itemSize !== 0) {
    throw new Error(`section "${name}": byteLength ${section.byteLength} is not a multiple of its ${section.dtype} element size`);
  }
  if (section.byteOffset + section.byteLength > buffer.byteLength) {
    throw new Error(`section "${name}": extends past the end of the binary file`);
  }
  const count = section.byteLength / itemSize;
  const expectedCount = section.shape.reduce((a, b) => a * b, 1);
  if (count !== expectedCount) {
    throw new Error(
      `section "${name}": byteLength implies ${count} element(s) but shape [${section.shape.join(",")}] implies ${expectedCount}`,
    );
  }

  if (PLATFORM_IS_LITTLE_ENDIAN) {
    return section.dtype === "float32"
      ? new Float32Array(buffer, section.byteOffset, count)
      : new Uint16Array(buffer, section.byteOffset, count);
  }
  const view = new DataView(buffer, section.byteOffset, section.byteLength);
  if (section.dtype === "float32") {
    const out = new Float32Array(count);
    for (let i = 0; i < count; i += 1) out[i] = view.getFloat32(i * 4, true);
    return out;
  }
  const out = new Uint16Array(count);
  for (let i = 0; i < count; i += 1) out[i] = view.getUint16(i * 2, true);
  return out;
}

export interface LoadedBinaryAsset<M extends BinaryManifestBase> {
  manifest: M;
  sections: Record<string, DecodedSection>;
}

export type LoadResult<M extends BinaryManifestBase> =
  | { ok: true; value: LoadedBinaryAsset<M> }
  | { ok: false; error: string };

function formatZodError(prefix: string, error: z.ZodError): string {
  const msg = error.issues
    .map((i) => {
      const p = i.path.join(".");
      return p ? `${p}: ${i.message}` : i.message;
    })
    .join("; ");
  return `${prefix}: ${msg}`;
}

/**
 * Validate a manifest against `manifestSchema`, fetch + validate its sibling
 * binary payload (byte length, and a SHA-256 digest when SubtleCrypto is
 * available), and decode every declared section into a typed array. Returns
 * a clear, student-friendly error string on any failure -- never throws.
 */
export async function loadBinaryAsset<M extends BinaryManifestBase>(
  manifestJson: unknown,
  manifestSchema: z.ZodType<M>,
  manifestUrl: URL,
): Promise<LoadResult<M>> {
  const parsed = manifestSchema.safeParse(manifestJson);
  if (!parsed.success) {
    return { ok: false, error: formatZodError("Invalid activity manifest", parsed.error) };
  }
  const manifest = parsed.data;

  for (const key of Object.keys(manifest)) {
    if (IDENTIFIER_TOKEN.test(key)) {
      return { ok: false, error: `identifier-shaped top-level key "${key}" is not allowed` };
    }
  }
  for (const key of Object.keys(manifest.sections)) {
    if (IDENTIFIER_TOKEN.test(key)) {
      return { ok: false, error: `identifier-shaped section name "${key}" is not allowed` };
    }
  }

  const binaryUrlResult = resolveDataUrl(manifestUrl.href, manifest.binary.path);
  if (!binaryUrlResult.ok) {
    return { ok: false, error: `Activity data file path is not safe to load: ${binaryUrlResult.error}` };
  }

  let buffer: ArrayBuffer;
  try {
    buffer = await fetchArrayBuffer(binaryUrlResult.url);
  } catch (e) {
    return { ok: false, error: `Could not load this activity's data file: ${(e as Error).message}` };
  }

  if (buffer.byteLength !== manifest.binary.byteLength) {
    return {
      ok: false,
      error:
        `This activity's data file looks incomplete or out of date ` +
        `(expected ${manifest.binary.byteLength} bytes, got ${buffer.byteLength}). Try reloading the page.`,
    };
  }

  const digest = await sha256Hex(buffer);
  if (digest !== null && digest.toLowerCase() !== manifest.binary.sha256.toLowerCase()) {
    return {
      ok: false,
      error: "This activity's data file failed an integrity check. Try reloading the page.",
    };
  }

  const sections: Record<string, DecodedSection> = {};
  try {
    for (const [name, section] of Object.entries(manifest.sections)) {
      sections[name] = decodeSection(buffer, name, section);
    }
  } catch (e) {
    return { ok: false, error: `This activity's data file is malformed: ${(e as Error).message}` };
  }

  return { ok: true, value: { manifest, sections } };
}

File: scripts/binary_asset.py
#!/usr/bin/env python3
"""Shared deterministic binary-payload packer/decoder for browser interactive
assets (WP15 §3).

Replaces large numeric-array JSON payloads with a small versioned JSON
manifest plus one deterministic binary file, read in the browser with
``fetch(...).arrayBuffer()`` and typed arrays (``interactive/src/binary-asset.ts``
is the matching TypeScript decoder -- keep both in sync: dtypes, alignment,
and endianness must match exactly).

Format:

* every section is a flat, C-contiguous array of one dtype (``float32`` or
  ``uint16``), always written little-endian regardless of the host machine;
* sections are concatenated back-to-back, each padded with zero bytes so the
  NEXT section starts at a 4-byte-aligned offset (so a ``Float32Array`` view
  can be constructed directly on the underlying ``ArrayBuffer`` without a
  copy, on every browser);
* the JSON manifest records, per section: ``dtype``, ``shape``, ``byteOffset``,
  ``byteLength`` -- everything the browser loader needs to slice + validate
  the buffer, plus the whole binary file's own ``byteLength`` and SHA-256
  digest for integrity verification.

``uint16`` is used for nearest-neighbour row indices; every caller must prove
(assert) the maximum index is below 65,536 before adding such a section --
:meth:`BinaryAssetBuilder.add` raises if it is not.
"""

from __future__ import annotations

import hashlib
from typing import Any

ALIGNMENT = 4
NUMPY_DTYPE = {"float32": "<f4", "uint16": "<u2"}
ELEMENT_SIZE = {"float32": 4, "uint16": 2}


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


class BinaryAssetBuilder:
    """Accumulates named sections into one deterministic, 4-byte-aligned,
    little-endian binary blob, and the matching manifest ``sections`` dict."""

    def __init__(self) -> None:
        self._chunks: list[bytes] = []
        self._offset = 0
        self.sections: dict[str, dict[str, Any]] = {}

    def add(self, name: str, dtype: str, array: Any) -> None:
        import numpy as np

        if name in self.sections:
            raise ValueError(f"duplicate section name {name!r}")
        if dtype not in NUMPY_DTYPE:
            raise ValueError(f"section {name!r}: unsupported dtype {dtype!r}; known: {sorted(NUMPY_DTYPE)}")
        arr = np.asarray(array)
        if dtype == "uint16":
            if arr.size and (int(arr.min()) < 0 or int(arr.max()) > 65535):
                raise ValueError(
                    f"section {name!r}: value(s) outside the uint16 range [0, 65535] "
                    f"(min={arr.min() if arr.size else None}, max={arr.max() if arr.size else None}); "
                    "this data no longer fits the 16-bit index format and needs a wider dtype."
                )
        cast = arr.astype(NUMPY_DTYPE[dtype])
        raw = cast.tobytes(order="C")
        pad = (-len(raw)) % ALIGNMENT
        byte_offset = self._offset
        self._chunks.append(raw + b"\x00" * pad)
        self._offset += len(raw) + pad
        self.sections[name] = {
            "dtype": dtype,
            "shape": list(cast.shape),
            "byteOffset": byte_offset,
            "byteLength": len(raw),
        }

    def build(self) -> bytes:
        return b"".join(self._chunks)


def decode_section(blob: bytes, section: dict[str, Any]) -> Any:
    """Reconstruct one section's numpy array from a built blob + its manifest
    entry -- the Python-side mirror of the TypeScript loader, used to
    validate a built/committed artifact by reconstructing the SAME logical
    arrays the old JSON format stored directly."""
    import numpy as np

    dtype = NUMPY_DTYPE[section["dtype"]]
    offset = section["byteOffset"]
    length = section["byteLength"]
    itemsize = ELEMENT_SIZE[section["dtype"]]
    if length % itemsize != 0:
        raise ValueError(f"section byteLength {length} is not a multiple of its dtype's element size {itemsize}")
    count = length // itemsize
    arr = np.frombuffer(blob, dtype=dtype, count=count, offset=offset)
    return arr.reshape(tuple(section["shape"]))
